fixes #16736 - UI Notifications support rails part

ohadlevy opened the pull request:
- [x] DB Model.
- [x] JSON API Endpoints.

UI etc will done in a separate PR.

ShimShtein commented:
I was thinking about usage API, and it led me to a question:
Why do we have to store notification types in database? IMHO we can easily get away with plain Ruby object that will describe notifications and will be stored in a hash. Something along the lines of:

``` ruby
@@notification_types = {
  :host_registered => NotificationType.new(audience: :user, :message 'New host registered %id', level: :info),
  :host_provisioned => NotificationType.new(audience: :user, :message 'Host %id successfully provisioned', level: :info)
}
```

Plugins could extend this list of course. The API would be pretty similar to regular logging statement:

``` ruby
#hosts_controller.rb
# ... snip ...
Notifications.emit(notification_type: :host_provisioned, initiator: current_user, subject: @host)
```

Advantages:
- easy to extend
- less queries to the DB
- Easy to use for coders

Disadvantages:
- Can be easily abused

lzap commented on app/controllers/user_notifications_controller.rb:
```diff
@@ -0,0 +1,25 @@
+    order(:created_at).
+    eager_load(:notification, :notification_type)
+
+    render :json => @notifications.map(&:payload)
```
No pagination will be problem very soon.

lzap commented on app/models/notification_type.rb:
```diff
@@ -0,0 +1,34 @@
+  AUDIENCE_GROUP    = 'usergroup'.freeze
+  AUDIENCE_TAXONOMY = 'taxonomy'.freeze
+  AUDIENCE_GLOBAL   = 'global'.freeze
+  AUDIENCE_ADMIN    = 'admin'.freeze
```
Isn't the `# frozen_string_literal: true` statement better option?

lzap commented on app/models/notification.rb:
```diff
@@ -0,0 +1,41 @@
+    # notification_recipients.each do |notice|
+    # ActionCable.server.broadcast("notifications_#{notice.user_id}", notice.payload)
+    # end
+  end
```
It is worth decoupling the ActionCable from this, so there can be different implementations - email, SMS, ActionCable.

ShimShtein replied:
Maybe just a callback, so it would be extendable by observers?

ohadlevy commented:
@iNecas - rebased, minor test failure atm will have a look later, but should not block you from looking at the pr :)

you can probably ignore the other commit around UI if it clutters your view.

dLobatog commented on app/models/notification.rb:
```diff
@@ -0,0 +1,41 @@
+
+  private
+
+  def calcuate_expiry
```
typo, calculate

dLobatog commented on app/models/notification.rb:
```diff
@@ -0,0 +1,41 @@
+  validates :notification_type, :presence => true
+  validates :initiator, :presence => true
+  validates :subject, :presence => true, :allow_nil => true
+  before_create :calcuate_expiry, :set_notification_recipients
```
typo, calculate

dLobatog commented on app/controllers/user_notifications_controller.rb:
```diff
@@ -0,0 +1,29 @@
+  end
+
+  def find_resource
+    @user_notification = NotificationRecipient.where(:user_id => User.current.id, :id => params[:id]).first
```
Doesn't look like User.current will work if `SETTINGS[:login] = false`, then again I'm really all for throwing away that and `:unattended`...

ohadlevy replied:
right.. I assume we should not enable notifications if settings[:login] is false?

dLobatog commented on app/models/notification.rb:
```diff
@@ -0,0 +1,41 @@
+  has_many :recipients, :class_name => User, :through => :notification_recipients, :source => :user
+
+  validates :notification_type, :presence => true
+  validates :initiator, :presence => true
```
Why is the `:initiator` needed when the audience isn't `user`?

ohadlevy replied:
it simply marks who initated the event that lead to the notification, think user X deleted a whole bunch of stuff and now you need to be aware of that.
or another usage case, your org admin (user x) created a new hostgroup that you can now use.

dLobatog replied:
OK - should it default to `User::ANONYMOUS_ADMIN`?

dLobatog commented on app/models/notification_type.rb:
```diff
@@ -0,0 +1,34 @@
+  def subscriber_ids(initiator, subject = nil)
+    case audience
+    when AUDIENCE_GLOBAL
+      User.reorder('').uniq.pluck(:id)
```
Why reorder? `.uniq` is supposed to be guaranteed by the db too

ohadlevy replied:
reorder is just faster (no need to sort by default scope - now or in the future), you are probably right about the uniq.

dLobatog commented on app/models/notification.rb:
```diff
@@ -0,0 +1,41 @@
+
+  validates :notification_type, :presence => true
+  validates :initiator, :presence => true
+  validates :subject, :presence => true, :allow_nil => true
```
What is subject required for when the audience is not `taxonomy` or `group`? Also it seems that `:presence => true, :allow_nil => true` contradict each other

ohadlevy replied:
the idea behind the subject is that you can refer to a specific obj, e.g. host X was deleted, in that case subject = host.

not every event has a subject (some subsystem might be down) hence the allow_nil, tbh I cant remember the line of thought around the two validations... :-\

dLobatog commented on db/migrate/20160927071039_create_notification_types.rb:
```diff
@@ -0,0 +1,13 @@
+class CreateNotificationTypes < ActiveRecord::Migration
+  def change
+    create_table :notification_types do |t|
+      t.string :name, index: true, unique: true
```
`:group` would be more fitting

ohadlevy replied:
should we add a name and a group? esp if its going to become a blueprint i would think:
name - provisioning error, group: host or provisioning

dLobatog replied:
From what I see on the UI side http://www.patternfly.org/pattern-library/communication/notification-drawer/#/api only the `group` and `message` are displayed, therefore the name would only be used internally, and the table already has an id

ohadlevy replied:
right, but name is uniq, group shouldn't be.

dLobatog commented on app/models/notification.rb:
```diff
@@ -0,0 +1,41 @@
+    self.notification_recipients.build subscribers.map{|id| { :user_id => id}}
+  end
+
+  def emit_message
```
Dead code so far, I think it should be removed if it's not doing anything.

dLobatog commented on app/models/notification.rb:
```diff
@@ -0,0 +1,41 @@
+  scope :expired, -> { where('expired_at < :now', {:now => Time.now.utc}) }
+
+  def type=(type)
+    self.notification_type = NotificationType.find_by_name!(type)
```
Why do you want to overwrite the setter and raise a ActiveRecord error when there's already a validation?

ohadlevy replied:
its just a short form, probably alias would be a better fit.

dLobatog reviewed (changes requested):
@ohadlevy I made some comments inline. Other than that, my biggest gripe is that NotificationType and Notification have intertwined dependencies to resolve the recipients. This is what I interpret from the code I read:

* NotificationType defines message, *audience*, level, expiration
* Notification defines initiator, subject and *recipients*. It defines the recipients by mixing data between the two classes, when it's not really necessary.

I think the intention of the classes here would be more clear by doing the following.

* NotificationType -> rename to NotificationTemplate or better, NotificationBlueprint (to not confuse it with other templates in Foreman). Move `subscriber_ids` and `audience` to `Notification`. This class responsibility is to be a blueprint to create notifications. It stores the message, level, and expiration.
* Notification - this class' responsibility is to define the recipients for the NotificationBlueprint. It merely takes a notification, and creates a bunch of NotificationRecipients to be picked up by the controller. 

Now, another thing I didn't understand well is the role of the `subject`, is it something static (e.g: a NotificationBlueprint for subscription expiration would always use the subject Katello::Subscription) or dynamic? (e.g: a NotificationBlueprint for subscription expiration might change it's subject?). To me it seems that it's something static, that depends on the content of the NotificationBlueprint. If so, why not moving it to NotificationBlueprint?  My intent again is to separate reponsibilities and move 'static' things to the blueprint.

I tested this mostly with https://github.com/gailsteiger/foreman/commits/n-ui and https://gist.github.com/dLobatog/3f4396229c7b59b06dd13c9102e6aee7, good job @gailsteiger & @ohadlevy , I think it's a solid model to move forward.

Let me know what you think, or if you want/need any help with this, I'm happy to make the changes on a separate branch and send them to you.

ohadlevy committed:
fixes #17545 - adds UI notification support

- Initial data model for notification support, based on manageiq design.
- addes notification JSON endpoints.

ShimShtein commented on app/controllers/user_notifications_controller.rb:
```diff
@@ -0,0 +1,36 @@
+    eager_load(:notification, :notification_type)
+
+    render :json => {
+      :notifications => @notifications.paginate(paginate_options).map(&:payload),
```
Why not use RABL to generate the JSON?

ohadlevy replied:
I didn't see much value in it atm?

ShimShtein commented on app/models/notification_recipient.rb:
```diff
@@ -0,0 +1,20 @@
+      :level      => notification_type.level,
+      :created_at => notification.created_at,
+      :text       => notification_type.message,
+      :seen       => seen
```
Worth adding `:subject` properties here. Otherwise, it's a bit useless on the UI side.

ohadlevy replied:
just adding subject would include a to_json on the object, I think we need a better way of how to pass the object (and its actions) to the UI.

ShimShtein replied:
Agreed. How about `:partial` approach? Using either a registry (class to set of properties) or in case of rabl, actual partial templates with proper naming convention?

ShimShtein commented on app/controllers/user_notifications_controller.rb:
```diff
@@ -0,0 +1,36 @@
+
+  def update
+    # only allowed attribute to change is 'seen'
+    process_response @user_notification.update_attribute(:seen, params[:user_notification][:seen])
```
Need to move to `Parameters` paradigm?

dLobatog replied:
I don't think this works already, as the params have not been permitted here. I'll fix that on my update. Thanks for the catch 😄

ShimShtein commented on app/controllers/user_notifications_controller.rb:
```diff
@@ -0,0 +1,36 @@
+    }
+  end
+
+  def update
```
I am missing "Mark all as read" action. IMHO it should be useful.
Other option would be bulk update action, It's much more efficient from DB point of view, and in the case of notifications IMHO we want to be as efficient as possible.

dLobatog replied:
Let's keep these for next PRs, I'd really prefer for this to not go over the current LOC - you can always submit a PR based on this branch as https://github.com/theforeman/foreman/pull/4170 did

dLobatog commented on app/controllers/user_notifications_controller.rb:
```diff
@@ -0,0 +1,36 @@
+  end
+
+  def find_resource
+    @user_notification = NotificationRecipient.where(:user_id => User.current.id, :id => params[:id]).first
```
This seems like a hint that `NotificationRecipient` should really be `UserNotification`, I'll see if I can find a way to call find_resource

dLobatog commented:
https://gist.github.com/dLobatog/76071b49bfc3f32f9a70f1f16546fbc9 to test this

dLobatog commented:
https://github.com/theforeman/foreman/pull/4194/commits follows  the changes I was requesting - I have to test it more thoroughly though

ohadlevy commented:
replaced by #4194
